### PR TITLE
Add Moving Least Squares example

### PR DESCRIPTION
**Pull request by mrlag31**
This pull request adds a new example for ArborX, showing how it can be used to compute and approximate a function from a point cloud to another, using the Moving Least Squares method. This example is taken from [DataTransferKit's MeshFree](https://github.com/ORNL-CEES/DataTransferKit/tree/master/packages/Meshfree/src) package and removes all usage of DTK. This is a draft PR as some extra work is still needed.

Some important differences from DTK's code:
- Views are multidimensional instead of data vectors. I feel like it helps better understand the layout of each object and the operations done on them.
- The inverse is computed using Gaussian elimination instead of SVD as a start.
- Data is baked within the example.

### CODE REVIEW

File: examples/moving_least_squares/moving_least_squares.cpp [hunk 1]
```diff
@@ -0,0 +1,335 @@
+/****************************************************************************
+ * Copyright (c) 2017-2023 by the ArborX authors                            *
```
Comment (aprokop) [inline examples/moving_least_squares/moving_least_squares.cpp:2]:
```suggestion
 * Copyright (c) 2023 by the ArborX authors                                 *
```

File: examples/moving_least_squares/moving_least_squares.cpp [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+// Example taken from DataTransferKit
+// (https://github.com/ORNL-CEES/DataTransferKit)
+// with MLS resolution from
+// (http://dx.doi.org/10.1016/j.jcp.2015.11.055)
```
Comment (aprokop) [inline examples/moving_least_squares/moving_least_squares.cpp:15]: Not sure if we need this.

File: examples/moving_least_squares/moving_least_squares.cpp [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+  double _radius;
+};
+
+struct MVPolynomialBasis_3D
+{
+  static constexpr std::size_t size = 4;
+
+  template <typename Double3D>
+  KOKKOS_INLINE_FUNCTION Kokkos::Array<double, size>
+  operator()(Double3D const &p) const
```
Comment (aprokop) [inline examples/moving_least_squares/moving_least_squares.cpp:42]: I think it should be templated on a `Point`. We can assert that the point is 3D (can we extend the basis to other dimensions)? Also, would want to switch away from using `double`.

File: examples/moving_least_squares/moving_least_squares.cpp [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+  {
+    return {{1., p[0], p[1], p[2]}};
+  }
+};
+
+// Func to evaluate
+template <typename Double3D>
+KOKKOS_INLINE_FUNCTION double func(Double3D const &p) {
```
Comment (aprokop) [inline examples/moving_least_squares/moving_least_squares.cpp:50]: Maybe a more meaningful name for a function, possibly `manufactured_solution`?

File: examples/moving_least_squares/moving_least_squares.cpp [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+
+int main(int argc, char *argv[])
+{
+  Kokkos::ScopeGuard guard(argc, argv);
+  constexpr std::size_t num_neighbors = 10;
+  constexpr std::size_t source_points_num = 10;
+  constexpr std::size_t target_points_num = 4;
+
+  auto source_points = Kokkos::View<ArborX::Point *, MemorySpace>(
+    "source_points", source_points_num);
```
Comment (aprokop) [inline examples/moving_least_squares/moving_least_squares.cpp:62]: We would want to prefix the names by something to indicate their origin. In examples, we would want them to be something like `Example::source_points`. This concerns all views created in the example.
 
You also don't need to initialize the points here, and can use `Kokkos::view_alloc` with `Kokkos::WithoutInitializing`, similar to how we do in other places.

File: examples/moving_least_squares/moving_least_squares.cpp [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+  auto target_points = Kokkos::View<ArborX::Point *, MemorySpace>(
+    "target_points", target_points_num);
+  auto source_points_host = Kokkos::create_mirror_view(source_points);
+  auto target_points_host = Kokkos::create_mirror_view(target_points);
+
+  // Generate source points
+  source_points_host(0) = ArborX::Point {  1.,  1.,  0. };
```
Comment (aprokop) [inline examples/moving_least_squares/moving_least_squares.cpp:69]: You have implicit conversions from `double` to float` here. Should probably be either `1` and `0`, or `1.f` and `0.f`.

File: examples/moving_least_squares/moving_least_squares.cpp [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+  constexpr std::size_t target_points_num = 4;
+
+  auto source_points = Kokkos::View<ArborX::Point *, MemorySpace>(
+    "source_points", source_points_num);
+  auto target_points = Kokkos::View<ArborX::Point *, MemorySpace>(
+    "target_points", target_points_num);
+  auto source_points_host = Kokkos::create_mirror_view(source_points);
+  auto target_points_host = Kokkos::create_mirror_view(target_points);
+
+  // Generate source points
```
Comment (aprokop) [inline examples/moving_least_squares/moving_least_squares.cpp:68]: It would help to provide a description or an ascii image of how the points are distributed.

File: examples/moving_least_squares/moving_least_squares.cpp [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+  source_points_host(1) = ArborX::Point { -1.,  1.,  0. };
+  source_points_host(2) = ArborX::Point { -1., -1.,  0. };
+  source_points_host(3) = ArborX::Point {  1., -1.,  0. };
+  source_points_host(4) = ArborX::Point {  0.,  0.,  1. };
+  source_points_host(5) = ArborX::Point {  1.,  1.,  2. };
+  source_points_host(6) = ArborX::Point { -1.,  1.,  2. };
+  source_points_host(7) = ArborX::Point { -1., -1.,  2. };
+  source_points_host(8) = ArborX::Point {  1., -1.,  2. };
+  source_points_host(9) = ArborX::Point {  0.,  0., -1. };
+  Kokkos::deep_copy(source_points, source_points_host);
```
Comment (aprokop) [inline examples/moving_least_squares/moving_least_squares.cpp:79]: We typically want to use `deep_copy` with a provided execution space.

File: examples/moving_least_squares/moving_least_squares.cpp [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+  // Generate target points
+  target_points_host(0) = ArborX::Point {  0.,  0.,  0. };
+  target_points_host(1) = ArborX::Point {  .5,  .5,  0. };
+  target_points_host(2) = ArborX::Point { -.5,  .5,  1. };
+  target_points_host(3) = ArborX::Point {  0., -.5, 1.5 };
+  Kokkos::deep_copy(target_points, target_points_host);
+
+  // Arrange source points as tree
+  auto source_tree =
+    ArborX::BVH<MemorySpace>(ExecutionSpace{}, source_points);
```
Comment (aprokop) [inline examples/moving_least_squares/moving_least_squares.cpp:90]:
```suggestion
  ArborX::BVH<MemorySpace> source_tree(ExecutionSpace{}, source_points);
```

File: examples/moving_least_squares/moving_least_squares.cpp [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+          source_points(indices(j))[1] - target_points(i)[1],
+          source_points(indices(j))[2] - target_points(i)[2],
+        };
+      }
+  });
+
+  // Compute the radii for the weight (phi) vector
+  auto radii = Kokkos::View<double*, MemorySpace>(
+    "radii", target_points_num);
+  constexpr double epsilon = std::numeric_limits<double>::epsilon();
```
Comment (aprokop) [inline examples/moving_least_squares/moving_least_squares.cpp:129]: Probably want `float` here, it's up for a discussion.

File: examples/moving_least_squares/moving_least_squares.cpp [hunk 11]
```diff
@@ -0,0 +1,335 @@
...
+      }
+
+      radii(i) = 1.1 * radius;
+  });
+
+  // Compute the weight (phi) vector
+  auto phi = Kokkos::View<double**, MemorySpace>(
+    "phi", target_points_num, num_neighbors);
+  Kokkos::parallel_for(
+    "phi_computation",
```
Comment (aprokop) [inline examples/moving_least_squares/moving_least_squares.cpp:150]: We also try to prefix the kernels to indicate the origin, like `Example::phi_computation`.

File: examples/moving_least_squares/moving_least_squares.cpp [hunk 12]
```diff
@@ -0,0 +1,335 @@
...
+  auto target_values_host = Kokkos::create_mirror_view(target_values);
+  Kokkos::deep_copy(target_values_host, target_values);
+  auto target_values_exact_host = Kokkos::create_mirror_view(target_values_exact);
+  Kokkos::deep_copy(target_values_exact_host, target_values_exact);
+
+  for (int i = 0; i < target_points_num; i++) {
+    std::cout << "====\n"
+              << target_values_host(i) << '\n'
+              << target_values_exact_host(i) << "\n====\n";
+  }
```
Comment (aprokop) [inline examples/moving_least_squares/moving_least_squares.cpp:334]: Probably want to compute and print an error.

**Review by aprokop — COMMENTED** [review]
Really good start. Quick partial review to address some of the common issues.

File: examples/moving_least_squares/moving_least_squares.cpp (hunk 2 above)
Comment (mrlag31) [reply examples/moving_least_squares/moving_least_squares.cpp:15]: This is mostly for me, if I either have to write a report or extend the comments to explain the methods. These are most likely temporary.

File: examples/moving_least_squares/moving_least_squares.cpp (hunk 3 above)
Comment (mrlag31) [reply examples/moving_least_squares/moving_least_squares.cpp:42]: This was taken directly from DTK's code of different polynomial bases. I will strip out the templates and switch `double` to `float` on the example. As I am using `ArborX::Point`, it is 3D.

I think it is possible to generate that code for any number of dimensions and polynomial degree, but it will most likely require a lot of templated code and classes.

File: examples/moving_least_squares/moving_least_squares.cpp (hunk 7 above)
Comment (mrlag31) [reply examples/moving_least_squares/moving_least_squares.cpp:68]: I will change how points are generated as this was only the basis to have the minimum code required to be able to test the example. I'll probably generate them to be as a mesh inside a cube.

File: examples/moving_least_squares/moving_least_squares.cpp (hunk 10 above)
Comment (mrlag31) [reply examples/moving_least_squares/moving_least_squares.cpp:129]: It depends a lot on how the values should be computed. Should the coefficients be handled as doubles or floats?

Comment (aprokop) [reply examples/moving_least_squares/moving_least_squares.cpp:129]: It may make sense to do the calculations based on the floating type of the data. We have `Point<DIM, FloatingPoint>`, so should probably use the same type as the second template argument of the `Point`. Note that we need #918 with `GeometricTraits::coordinate_type<Point>::type` to figure out that type.

**Comment by aprokop** [comment]
add to whitelist

**Comment by aprokop** [comment]
ok to test

**Comment by aprokop** [comment]
@mrlag31 You'll need to use clang-format 14 and fix the styling (e.g., using `scripts/check_format_cpp.sh`) to be able to test the PR.

**Comment by mrlag31** [comment]
I've fixed some of the issues you found.
- No templates (replaced with `Point`)
- Now using only `float`s
- Proper source point generation (but should also be done for target points)

File: examples/moving_least_squares/moving_least_squares.cpp [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+  // Create the queries
+  // For each target point we query the closest source points
+  auto queries = Kokkos::View<ArborX::Nearest<ArborX::Point> *, MemorySpace>(
+      "MLS_EX::queries", target_points_num);
+  Kokkos::parallel_for(
+      "MLS_EX::make_queries",
+      Kokkos::RangePolicy<ExecutionSpace>(0, target_points_num),
+      KOKKOS_LAMBDA(int const i) {
+        queries(i) = ArborX::nearest(target_points(i), num_neighbors);
+      });
```
[the remark this answers is not in the document]
Comment (mrlag31) [reply examples/moving_least_squares/moving_least_squares.cpp:101]: That's how it was done in DTK, so I copied. But I agree that using `AccessTraits` would be a more "correct" way to do it.

File: examples/moving_least_squares/moving_least_squares.cpp [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+  auto radii =
+      Kokkos::View<float *, MemorySpace>("MLS_EX::radii", target_points_num);
+  constexpr float epsilon = std::numeric_limits<float>::epsilon();
+  Kokkos::parallel_for(
+      "MLS_EX::radii_computation",
+      Kokkos::RangePolicy<ExecutionSpace>(0, target_points_num),
+      KOKKOS_LAMBDA(int const i) {
+        float radius = 10. * epsilon;
+
+        for (int j = 0; j < num_neighbors; j++)
```
[the remark this answers is not in the document]
Comment (mrlag31) [reply examples/moving_least_squares/moving_least_squares.cpp:137]: If the number of source points is less than the number of requested neighbors, I would most likely consider that an input error. Is there other reasons on why would a query not return the requested number of neighbors?

File: examples/moving_least_squares/moving_least_squares.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (aprokop) [reply examples/moving_least_squares/moving_least_squares.cpp:101]: > That's how it was done in DTK

We've come a long way since DTK. DTK is pretty outdated in the way it uses ArborX. It is a good starting point, but we want and should move more towards a better and more efficient use of ArborX.

File: examples/moving_least_squares/moving_least_squares.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (aprokop) [reply examples/moving_least_squares/moving_least_squares.cpp:137]: Yes, it is the only case where it would be not `num_neighbors`.

**Comment by mrlag31** [comment]
I have switched from using Gaussian method to the SVD pseudo-inverse. I have also made adjustments from DTK's version by using the fact that the moment matrix is symmetric.

**Review by aprokop — COMMENTED** [review]
I think DTK has a pretty decent interface for `MLS`, and we should follow it more closely here, with some differences:
- Switch `DeviceType` to `MemorySpace` (do we need it?)
- Introduce `ExecutionSpace` to the constructor and `apply`
- Duplicate communicator (similar to how it's done in the `DistributedTree`)
- (maybe) store `Distributor`

Other things:
- Anything regarding data setup should probably belong to the example `.cpp` (e.g., `TargetPoints`)
- I think `MLSComputation` is really a collection of free functions, which should be in `ArborX_DetailsMovingLeastSquares.hpp` in the `Details` namespace (similar to DTK)
- Likely don't want to hardcode user data to be Views (or, at least, point data), and switch to using `AccessTraits` to access source and target points
- We prefer function to follow camel case (and other styling things, see [here](https://github.com/arborx/ArborX/wiki/CodeStyle))
- `num_neighbors` should not be a parameter, it is determined from the polynomial basis size

File: examples/moving_least_squares/mls.hpp [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+    Kokkos::View<int *, MemorySpace> offsets("Example::MLS::offsets", 0);
+    source_tree.query(
+        space, TargetPoints<MemorySpace, Points>{target_points, _num_neighbors},
+        index_ranks, offsets);
+
+    // Split indices/ranks
+    Kokkos::View<int *, MemorySpace> local_indices(
+        "Example::MLS::local_indices", _tgt_size * _num_neighbors);
+    Kokkos::View<int *, MemorySpace> local_ranks("Example::MLS::local_ranks",
+                                                 _tgt_size * _num_neighbors);
```
Comment (aprokop) [inline examples/moving_least_squares/mls.hpp:83]: These views don't need initializations, as they will be set in the parallel loop below. So, please use `WithoutInitializing`.

File: examples/moving_least_squares/mls.hpp [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+  {
+    // There must be enough source points
+    assert(_src_size >= _num_neighbors);
+
+    // Organize source points as tree
+    ArborX::DistributedTree<MemorySpace> source_tree(comm, space,
+                                                     source_points);
+
+    // Perform the query
+    Kokkos::View<Kokkos::pair<int, int> *, MemorySpace> index_ranks(
```
Comment (aprokop) [inline examples/moving_least_squares/mls.hpp:72]: We have a type for that, I think, `PairIndexRank` from `src/details/ArborX_PairIndexRank.hpp`.

File: examples/moving_least_squares/symmetric_pseudoinverse_svd.hpp [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+#include <cmath>
+#include <limits>
+
+// Pseudo-inverse moment matrix using SVD
+// We must find U, E (diagonal and positive) and V such that A = U.E.V^T
+// We also know that A is symmetric (by construction), so U = SV where S is
+// a sign matrix (only 1 or -1 in the diagonal, 0 elsewhere).
+// Thus A = U.E.S.U^T
+template <class ValueType, typename MemorySpace>
+class SymmPseudoInverseSVD
```
Comment (aprokop) [inline examples/moving_least_squares/symmetric_pseudoinverse_svd.hpp:26]: Maybe we should move this in a separate PR, and then add some tests for it?

Comment (aprokop) [reply examples/moving_least_squares/symmetric_pseudoinverse_svd.hpp:26]: I'm also a bit confused. Why is this a class, and why do we need to store data? It's called exactly once from MLS, and I don't think we use any of the stored `_es`, `_inv`, etc, anywhere in MLS. Why not make it a free function?

File: examples/moving_least_squares/mpi_comms.hpp [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+
+#include <cassert>
+#include <memory>
+#include <optional>
+
+#include "common.hpp"
+#include <mpi.h>
+
+template <typename MemorySpace>
+class MPIComms
```
Comment (aprokop) [inline examples/moving_least_squares/mpi_comms.hpp:26]: I haven't closely looked at this yet.

**Review by aprokop — COMMENTED** [review]
Another round of comments I had. Still haven't looked through all the code, though.

File: examples/moving_least_squares/mls.hpp [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+                                  std::size_t i)
+  {
+    return ArborX::nearest(tp.target_points(i), tp.num_neighbors);
+  }
+
+  using memory_space = MemorySpace;
+};
+
+template <typename ValueType, typename PolynomialBasis, typename RBF,
+          typename MemorySpace>
```
[the remark this answers is not in the document]
Comment (mrlag31) [reply examples/moving_least_squares/mls.hpp:50]: `ValueType` (and `FloatingPoint` in your example) is the type of the calculations and the stored matrices, but it might be different from the coordinate type and the evaluation type. `FloatingCalculationType` would be a better name?

File: examples/moving_least_squares/mls.hpp [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+class MLS
+{
+public:
+  template <typename ExecutionSpace, typename Points>
+  MLS(ExecutionSpace const &space, MPI_Comm comm, Points const &source_points,
+      Points const &target_points,
```
[the remark this answers is not in the document]
Comment (mrlag31) [reply examples/moving_least_squares/mls.hpp:56]: Both the `source_points` and `target_points` are coming from user space and represent array of points. But I guess they can still have different types, so I will probably add more types to the templat.

File: examples/moving_least_squares/symmetric_pseudoinverse_svd.hpp (hunk 17 above)
Comment (mrlag31) [reply examples/moving_least_squares/symmetric_pseudoinverse_svd.hpp:26]: > Maybe we should move this in a separate PR, and then add some tests for it?
Why not, as it is one of the more complex set of functions and could actually be used for other things.

> I'm also a bit confused. Why is this a class, and why do we need to store data? It's called exactly once from MLS, and I don't think we use any of the stored _es, _inv, etc, anywhere in MLS. Why not make it a free function?
This is usually how I create complex algorithms that need temporary variables. The user cannot instantiate it so these variables are dropped as soon as the static method returns. However, I can transform it so it uses only free functions.

File: examples/moving_least_squares/mls.hpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (aprokop) [reply examples/moving_least_squares/mls.hpp:49]: > `FloatingCalculationType` would be a better name?

May be a bit better than a generic `ValueType`, but it still does not quite express the intent. There must be a better name.

File: examples/moving_least_squares/symmetric_pseudoinverse_svd.hpp (hunk 17 above)
Comment (mrlag31) [reply examples/moving_least_squares/symmetric_pseudoinverse_svd.hpp]: I have modified this part so it resembles what one could find in `ArborX::Details`. The inverse is now computed using only free functions and with less restrictions on the input.

**Comment by masterleinad** [comment]
I'm concerned that this is getting too large. I would rather have a node-level (non-MPI) version and the MPI examples https://github.com/arborx/ArborX/pull/732 separately and possibly an extension using concepts in common with #737.

**Comment by mrlag31** [comment]
I have attempted to add support for hypergeometry (i.e. points that are not 3D) but `DistributedTree` does not support geometries that are not 3D. This seems to be dealt in PR #907.

**Comment by masterleinad** [comment]
> I have attempted to add support for hypergeometry (i.e. points that are not 3D) but `DistributedTree` does not support geometries that are not 3D. This seems to be dealt in PR #907 so I won't be able to test them with something else other than 3D points.

I would just not use `ExperimentalHyperGeometry::Point` then. Let's not make this pull request even more complex by suggesting to use functionality that is not available.

**Comment by aprokop** [comment]
> I would just not use `ExperimentalHyperGeometry::Point` then. Let's not make this pull request even more complex by suggesting to use functionality that is not available.

I think I suggested it before MPI was added.

**Comment by mrlag31** [comment]
> I'm concerned that this is getting too large. I would rather have a node-level (non-MPI) version and the MPI examples #732 separately and possibly an extension using concepts in common with #737.

It is true that this starts to be quite a large example (although it can be used as an ArborX add-on). I think it would be interesting to discuss how can this PR be split.

**Comment by aprokop** [comment]
> I think it would be interesting to discuss how can this PR be split.

One of the ways we could proceed (requires general consensus) is like this:
- Create `arborx/interpolation` directory
- Move SVD to `arborx/interpolation/src/details/ArborX_DetailsSVD.hpp`, and add testing to `arborx/interpolation/test`
- Add radial basis functions and tests
- Add non-distributed MLS and tests
- Add non-distributed MLS example to `arborx/interpolation/examples`
- Add distributed MLS and tests
- Add distributed MLS example

This is just an idea. It really depends on whether we want to grow the set of the interpolation algorithms in the future and make them easily available to a user.

**Comment by Rombur** [comment]
> It really depends on whether we want to grow the set of the interpolation algorithms in the future and make them easily available to a user.

Next year I have a project where I will need to couple two codes together (I don't  remember the name of the codes), so I may use this.

**Comment by aprokop** [comment]
Can this be closed?
